pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskGenerator.java

```diff
@@ -0,0 +1,419 @@
+      // huge payload to pinot-server in request. Batching the requests will help in reducing the payload size.
+      int numSegmentsBatchPerServerRequest = Integer.parseInt(
+          taskConfigs.getOrDefault(MinionConstants.UpsertCompactMergeTask.NUM_SEGMENTS_BATCH_PER_SERVER_REQUEST,
+              String.valueOf(DEFAULT_NUM_SEGMENTS_BATCH_PER_SERVER_REQUEST)));
+
+      Map<String, List<ValidDocIdsMetadataInfo>> validDocIdsMetadataList =
+          serverSegmentMetadataReader.getSegmentToValidDocIdsMetadataFromServer(tableNameWithType, serverToSegments,
+              serverToEndpoints, null, 60_000, ValidDocIdsType.SNAPSHOT.toString(),
+              numSegmentsBatchPerServerRequest);
+
+      Map<String, SegmentZKMetadata> candidateSegmentsMap =
+          candidateSegments.stream().collect(Collectors.toMap(SegmentZKMetadata::getSegmentName, Function.identity()));
+
+      List<String> alreadyMergedSegments = getAlreadyMergedSegments(allSegments);
+
+      SegmentSelectionResult segmentSelectionResult =
+          processValidDocIdsMetadata(taskConfigs, candidateSegmentsMap, validDocIdsMetadataList, alreadyMergedSegments);
+
+      if (!segmentSelectionResult.getSegmentsForDeletion().isEmpty()) {
+        pinotHelixResourceManager.deleteSegments(tableNameWithType, segmentSelectionResult.getSegmentsForDeletion(),
+            "0d");
+        LOGGER.info(
+            "Deleted segments containing only invalid records for table: {}, number of segments to be deleted: {}",
+            tableNameWithType, segmentSelectionResult.getSegmentsForDeletion());
+      }
+
+      int numTasks = 0;
+      int maxTasks = Integer.parseInt(taskConfigs.getOrDefault(MinionConstants.TABLE_MAX_NUM_TASKS_KEY,
+          String.valueOf(MinionConstants.DEFAULT_TABLE_MAX_NUM_TASKS)));
+      for (Map.Entry<Integer, List<List<SegmentMergerMetadata>>> entry
+          : segmentSelectionResult.getSegmentsForCompactMerge().entrySet()) {
+        if (numTasks == maxTasks) {
+          break;
+        }
+        List<List<SegmentMergerMetadata>> groups = entry.getValue();
+        // no valid groups found in the partition to merge
+        if (groups.isEmpty()) {
+          continue;
+        }
+        // there are no groups with more than 1 segment to merge
+        // this can be later removed if we want to just do single-segment compaction from this task
+        if (groups.get(0).size() <= 1) {
+          continue;
+        }
+        // TODO see if multiple groups of same partition can be added
+        Map<String, String> configs = new HashMap<>(getBaseTaskConfigs(tableConfig,
+            groups.get(0).stream()
+                .map(x -> x.getSegmentZKMetadata().getSegmentName()).collect(Collectors.toList())));
+        configs.put(MinionConstants.DOWNLOAD_URL_KEY,
+            StringUtils.join(groups.get(0).stream().map(x -> x.getSegmentZKMetadata().getDownloadUrl())
```

Comment:
Let's move the StringUtils stuff to a separate function? Not only improves readability but you can also write a quick UT. Same for `ORIGINAL_SEGMENT_CRC_KEY`